Show a logging error instead a raise for empty data reader retrival

tovaz92 opened the pull request:
Fix: change #2141 by partially reverting the changes done. Using a `error` logging insead a `raise`.

## Issues closed by this pull request:

Close #2168

jhardenberg commented on src/aqua/reader/reader.py:
```diff
@@ -374,7 +374,8 @@ def retrieve(self, var=None, level=None,
         else:
             if data is None or len(data.data_vars) == 0:
-                raise NoDataError("Retrieved empty dataset. Check varname used as first step.")
+                self.logger.error("Retrieved empty dataset. Check varname used as first step.")
```
Do you still need the `
from aqua.exceptions import NoDataError`
at the top?

tovaz92 commented:
[Details/Discussion on commit [26770fb](https://github.com/DestinE-Climate-DT/AQUA/pull/2170/commits/26770fb8634a34d72e8a34040614a1dc7424fc83)]:

In `reader.py` I removed the `return None`:
```python 
if data is None or len(data.data_vars) == 0:
                self.logger.error(f"Retrieved empty dataset for {var=}. First, check its existence in the data catalog.")
```
and added an extra check on time coord:
```python

            if 'time' in data.coords:
                if startdate and enddate and not ffdb:  # do not select if data come from FDB (already done)
                    data = data.sel(time=slice(startdate, enddate))
            else:
                self.logger.debug(f"Variable {var} has no time coordinate, skipping time selection.")
```

while in `diagnostic.py` I added a check:
```python 
        if hasattr(data, 'data_vars') and len(data.data_vars) == 0:
            logger = log_configure(log_name='Diagnostic_reader', log_level=loglevel)
            logger.error(f"Variable '{var}' contains no data for {model} {exp} {source}.")
```

- The reason to remove the `return None` in `reader.py` is to avoid a later failure in `diagnostic.py` at:

```python
# If the data is empty, raise an error
        if not data:
            raise ValueError(f"No data found for {model} {exp} {source} with variable {var}")
```
as missing data would set data=None, which would raise an error and make fail the aqua-analysis workflow I suspect.

@jhardenberg, @mnurisso what are your impression on these changes?

mnurisso commented on src/aqua/reader/reader.py:
```diff
@@ -374,10 +374,13 @@ def retrieve(self, var=None, level=None,
 
-            if startdate and enddate and not ffdb:  # do not select if data come from FDB (already done)
-                data = data.sel(time=slice(startdate, enddate))
+            if 'time' in data.coords:
```
Access to data.coords will trigger a raise if data is None.

tovaz92 replied:
the `if 'time' ...` is removed.

mnurisso commented:
I don't agree with some change:
- The time if is not fail-proof as explained in the comment above, I'm not against having it, but as it is now it will create bugs. The if has to not fail if `data` are `None` object. Btw, this extra if you created needs then a test, otherwise we will never monitor that case. I don't think btw we have so far any dataset with no time available in tests, so that it is not so easy to test.
- The error raised in the diagnostic is done on purpose. If you're retrieving one variable and you obtain a None, this is definitely an issue for the diagnostic and the class should raise an error because the plot cannot be done. It will be the CLI to have a try-except and eventually skip that plot and move to the following if data are missing. I'd then remove the modification in the `diagnostics.py`.

tovaz92 commented on src/aqua_diagnostics/core/diagnostic.py:
```diff
@@ -130,6 +130,10 @@ def _retrieve(model: str, exp: str, source: str, var: str = None, catalog: str =
 
         data = reader.retrieve(var=var)
 
+        if hasattr(data, 'data_vars') and len(data.data_vars) == 0:
```
When retriving a wrong variable it happen that I did not get `None` as type but `data` will not have any `data_vars` in it (i.e. empty dataset). The check are similar but not exactly the same. If the utility is too strict we can remove though.

tovaz92 commented on src/aqua/reader/reader.py:
```diff
@@ -374,10 +374,10 @@ def retrieve(self, var=None, level=None,
         else:
             if data is None or len(data.data_vars) == 0:
-                raise NoDataError("Retrieved empty dataset. Check varname used as first step.")
+                self.logger.error(f"Retrieved empty dataset for {var=}. First, check its existence in the data catalog.")
```
It is a pythonic way that will print " ... var=<varname> ... " in the string. I choose this formatting to pinpoint to that the variable is named `var` in the code.
I am ok also with the other way of writing it though.

tovaz92 commented:
If the tests are ok, I would merge.
